add support for exec

TomasTomecek opened the pull request:
TODO:
- [x] tests
- [x] don't print text from `stdin` to `stdout`
- [x] disable line-buffering (enables completion and terminal escape characters)
- [x] polish descriptions in new exec tests

~~You need patch from this PR https://github.com/docker/docker-py/pull/853 to make this work.~~

You need docker-py master.

Fixes: #30

dnephin commented on dockerpty/pty.py:
```diff
@@ -249,3 +249,103 @@ def _hijack_tty(self, pumps):
+        super(ExecPTY, self).__init__(client, container, interactive=interactive,
+                                      stdout=stdout, stderr=stderr, stdin=stdin)
+
+        self.exec_id = None
```
I'd like to avoid the inheritance if possible.  Maybe a bunch of this can be added to the `PseudoTerminal` class directly?  At least the args are the same.

TomasTomecek replied:
Yes, it's possible to merge the class into `PseudoTerminal`. The reason I subclassed is that `PseudoTerminal` class is tied to running a container and attaching to it. `exec` is different operation with different workflow and API calls. If we merge those two use cases together, it will be pretty messy.

dnephin commented on dockerpty/pty.py:
```diff
@@ -249,3 +249,103 @@ def _hijack_tty(self, pumps):
+
+    def exec_create(self, command):
+        self.exec_id = self.client.exec_create(self.container, command, tty=self.interactive, stdin=self.interactive)
+        return self.exec_id
```
I think `exec_create()` could happen outside of this class, maybe as part of `exec_command()` ?

Then the id can be passed into the `exec_command()` call, and we won't need to store it on self.

TomasTomecek replied:
I wasn't sure if it should be stateless or not so I made it as generic as possible. Will rewrite then.

dnephin commented on dockerpty/pty.py:
```diff
@@ -249,3 +249,103 @@ def _hijack_tty(self, pumps):
+        try:
+            is_tty_on = self.exec_info()["ProcessConfig"]["tty"]
+        except (KeyError, AttributeError):
+            is_tty_on = False
```
I think this check is basically the equivalent of the `is_raw()` call, no?

TomasTomecek replied:
seems so, except that path within metadata is different

dnephin commented on dockerpty/pty.py:
```diff
@@ -249,3 +249,103 @@ def _hijack_tty(self, pumps):
+        if size is not None:
+            rows, cols = size
+            try:
+                self.client.exec_resize(self.exec_id, height=rows, width=cols)
```
Since this whole function only differs by this one function call, it would be good to just parametrize the function in the base class. That

TomasTomecek replied:
True

dnephin commented on dockerpty/pty.py:
```diff
@@ -249,3 +249,103 @@ def _hijack_tty(self, pumps):
+        is closed.
+        """
+        self.exec_create(command)
+        self.run_exec()
```
With the change I suggested about moving exec_id out, this would be part of the module level `exec_command()`

dnephin commented on dockerpty/pty.py:
```diff
@@ -249,3 +249,103 @@ def _hijack_tty(self, pumps):
+        finally:
+            if flags:
+                for (pump, flag) in zip(pumps, flags):
+                    io.set_blocking(pump, flag)
```
This should mostly be the same as `start()` in the base class, right? Other than the two lines here: https://github.com/d11wtq/dockerpty/blob/master/dockerpty/pty.py#L146-L147 are there any differences?

Could we parametrize that as well?

TomasTomecek replied:
Will try.

TomasTomecek commented:
@dnephin thank you for the feedback; will rewrite once we settle on subclass/merge

dnephin commented:
Instead of a merge, what if we try and do it with composition. I think we could create a new `Operation` class, that handles the difference between run and exec. In the case of exec, it can also hold the exec_id.

The interface would be something like:

```python

class Operation(object):

    def is_raw():
        ...

    def start():
        ...
```
(there might be other methods I've missed)

Then you could have a `RunOperation()` and an `ExecOperation()` to handle the differences.  

The PseudoTerminal class then just takes an operation and works for both.

If `create_exec()` happens outside of the class, the exec_id can be passed into the `ExecOperation()`.

TomasTomecek commented:
@dnephin so I removed `exec` interface and refactored current codebase to use the new `Operation` interface. Is this what you had in mind? As soon as we settle on the interface, I'll readd the `exec` support. Btw this breaks backwards compatibility of PseudoTTY's constructor. Not sure if it meant to be an API.

dnephin commented:
Nice, I think this looks great.

We do use the PseudoTTY constructor directly from compose. If we need to break compatibility I think we can bump the major version.

TomasTomecek commented:
Added exec support: tested locally and works nicely for me.

Regarding tests: behave tests are broken for me and IMO need to be fixed. Frankly, I am not interested in fixing those. (I would not mind creating tests for exec)

dnephin commented on .travis.yml:
```diff
@@ -14,35 +14,26 @@
-
-  # ---------------------------------------------------------------------------
-  chmod +x tests.sh
-  chmod +x run.sh
```
I guess the test failures are related to this change?  Was this not working anymore with the new travis setup?

dnephin replied:
I restart the last build of master to see what happens

TomasTomecek commented:
The reason I changed `.travis.yml` was that *all* behave tests were getting `Connection refused`:
https://travis-ci.org/d11wtq/dockerpty/jobs/93037271 

Also, tests are being run against ancient version of docker -- 1.3:
https://github.com/d11wtq/dockerpty/blob/master/.travis.yml#L40

I wanted to get some feedback, so I changed the tests to use docker from travis.

As I'm looking at the master build you rescheduled, results of behave tests are the same as results in this PR:
https://travis-ci.org/d11wtq/dockerpty/jobs/93328555
https://travis-ci.org/d11wtq/dockerpty/jobs/65955024

dnephin commented:
Yup, I think the changes to `.travis.yml` are good

TomasTomecek commented:
started working on tests

TomasTomecek commented:
All tests now pass. I had to disable test with `watch` b/c I had no idea how that test supposed to work.

@dnephin, @aanand, please take a look

From my PoV, it's ready for final review.

aanand commented:
Nice stuff. I think the last two commits can be squashed, and the commits concerned with changes to the code should be grouped together.

We should probably hold off on merging until there's a stable docker-py out with the required changes.

I also have no idea how that `watch` test is supposed to work.

dnephin commented on .travis.yml:
```diff
@@ -14,35 +14,26 @@
 
 install: pip install -r $REQUIREMENTS
-script: ./run.sh
+script: py.test -q tests && behave -c
```
If `run.sh` isn't being used any more, we should delete it

dnephin commented:
Looks good, once docker-py release is out, we should be good to merge

aanand commented:
Since we're waiting to merge the code changes, do you think we could get the test-related commits in a separate PR so we can merge them sooner?

TomasTomecek commented:
rebased, updated (docstrings, pep8)

TomasTomecek commented:
rebased

there is a docker-py-1.7 RC, can we finally merge this? I've already spent so much time on this and am tired of constant rebasing and fixing the tests (which they doesn't work of course and I have no idea why).

@dnephin @aanand

aanand commented on requirements-dev.txt:
```diff
@@ -1,4 +1,4 @@
-docker-py>=0.3.2
+docker-py>=1.7.0
```
1.7.0 isn't out yet - if you change this to `docker-py==1.7.0rc2` it should work. That seems to be why CI is currently failing.

TomasTomecek replied:
I was testing against engine 1.10 rc1 and it broke on my laptop so I pushed the `requirements.txt` change in a hurry and went away; was talking about these broken runs:

https://travis-ci.org/d11wtq/dockerpty/jobs/104060547
https://travis-ci.org/d11wtq/dockerpty/jobs/104052086
https://travis-ci.org/d11wtq/dockerpty/jobs/104084403 

I have a thought how to improve the suite so it's easier to debug.

TomasTomecek replied:
got it, will update soon

TomasTomecek commented:
If this won't work, I'm going to wash dishes to a fast food.

aanand commented:
Looks like the test introduced in https://github.com/d11wtq/dockerpty/pull/52/files is failing - could something have regressed?

TomasTomecek commented:
This is really odd. Trying
```
docker run --log-driver=none -ti --rm busybox sh
```

results into

```
/ # 
```

But when I read docs it says:

```
none 	Disables any logging for the container. docker logs won’t be available with this driver.
```

So I really have no idea here.

aanand commented:
Yep, that's expected - you can still attach, you just can't retrieve past logs.

```console
$ docker run --log-driver=none -d --name=no-logs busybox top
925c25391dae0115d7d58cf5ba135d685e37c9be2f1b1c56e0e5726c53046957
$ docker logs no-logs
"logs" command is supported only for "json-file" and "journald" logging drivers (got: none)
```

Finally got features running locally (#55), so I'm investigating this now.

aanand commented:
Right, this fixes it for me:

```diff
diff --git a/features/interactive_terminal.feature b/features/interactive_terminal.feature
index 9844dcd..1d0e82e 100644
--- a/features/interactive_terminal.feature
+++ b/features/interactive_terminal.feature
@@ -33,7 +33,7 @@ Feature: Attaching to an interactive terminal in a docker container
   Scenario: Starting the PTY againt container with disabled logging
     Given I am using a TTY
     And I run "/bin/sh" in a docker container with a PTY and disabled logging
-    When I start dockerpty
+    When I start dockerpty with disabled logging
     Then I will see the output
       """
       / #
diff --git a/features/steps/step_definitions.py b/features/steps/step_definitions.py
index a0cd0b1..0a24198 100644
--- a/features/steps/step_definitions.py
+++ b/features/steps/step_definitions.py
@@ -128,7 +128,12 @@ def step_impl(ctx):
 
 @when('I start dockerpty')
 def step_impl(ctx):
-    alloc_pty(ctx, dockerpty.start, ctx.client, ctx.container)
+    alloc_pty(ctx, dockerpty.start, ctx.client, ctx.container, logs=True)
+
+
+@when('I start dockerpty with disabled logging')
+def step_impl(ctx):
+    alloc_pty(ctx, dockerpty.start, ctx.client, ctx.container, logs=False)
 
 
 @when('I exec "{cmd}" in a running docker container')
```

I don't know why it's not failing on master :/

aanand commented:
Though "with disabled logging" isn't really correct language when referring to `dockerpty.start` - it should probably just say "with no logs".

TomasTomecek commented:
Thanks @aanand, very nice work! I've committed the patch and set you as author. It indeed makes the tests pass. I have also verified that we are not reintroducing #51 (code taken from [here](https://github.com/d11wtq/dockerpty/issues/51#issue-122931808)):

1. configured logging driver, `logs=False`:

   ```python
   import dockerpty
   import docker
   from docker.utils import kwargs_from_env
   
   client = docker.Client(**kwargs_from_env())
   container = client.create_container(
       image='busybox:latest',
       stdin_open=True,
       tty=True,
       command='/bin/sh',
       host_config={"LogConfig": {
           "Type": "none"
       }}
   )
   
   dockerpty.start(client, container['Id'], interactive=True, logs=False)
   ```
   
   ```
   $ python test.py
   / #                    
   ```
2. configured logging driver, `logs=True`:

   ```python
   import dockerpty
   import docker
   from docker.utils import kwargs_from_env
   
   client = docker.Client(**kwargs_from_env())
   container = client.create_container(
       image='busybox:latest',
       stdin_open=True,
       tty=True,
       command='/bin/sh',
       host_config={"LogConfig": {
           "Type": "none"
       }}
   )
   
   dockerpty.start(client, container['Id'], interactive=True, logs=True)
   ```
   
   ```
   $ python test.py
   Error attaching: loggingfactory: Failed to get logging factory: logger: no log driver named 'none' is registered
   
   $ docker run -ti --log-driver=none -d busybox sh
   6a8f28b83e65d3687a456a4b3fd7f14c9bf6c4a046f5e862289ea5317933c44c
   $ docker logs 6a8f28b83e65d3687a456a4b3fd7f14c9bf6c4a046f5e862289ea5317933c44c
   "logs" command is supported only for "json-file" and "journald" logging drivers (got: none)
   ```

aanand commented:
LGTM

TomasTomecek commented:
rebased (fingers crossed)

aanand commented:
I get some `behave` failures on OSX. Some are also present in master, so we can fix those separately; some are a result of re-using the `docker.Client` object in both the parent and child process. I've got a fix working, so I'll do a PR for that separately too.

LGTM

dnephin commented:
LGTM
